src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1762,31 +1901,107 @@ public sealed class One : Tests
                 public override void Filter(
                     DecisionDagBuilder builder,
                     BoundDagTest test,
+                    DagState state,
                     IValueSet? whenTrueValues,
                     IValueSet? whenFalseValues,
                     out Tests whenTrue,
                     out Tests whenFalse,
                     ref bool foundExplicitNullTest)
                 {
+                    if (Test is BoundDagEvaluation)
+                    {
+                        // there cannot be any implication for an evaluation node.
+                        whenTrue = whenFalse = this;
+                        return;
+                    }
+
                     builder.CheckConsistentDecision(
                         test: test,
                         other: Test,
+                        state: state,
                         whenTrueValues: whenTrueValues,
                         whenFalseValues: whenFalseValues,
                         syntax: test.Syntax,
+                        precondition: out Tests precondition,
+                        sideeffect: out Tests sideeffect,
                         trueTestPermitsTrueOther: out bool trueDecisionPermitsTrueOther,
                         falseTestPermitsTrueOther: out bool falseDecisionPermitsTrueOther,
                         trueTestImpliesTrueOther: out bool trueDecisionImpliesTrueOther,
                         falseTestImpliesTrueOther: out bool falseDecisionImpliesTrueOther,
                         foundExplicitNullTest: ref foundExplicitNullTest);
-                    whenTrue = trueDecisionImpliesTrueOther ? Tests.True.Instance : trueDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
-                    whenFalse = falseDecisionImpliesTrueOther ? Tests.True.Instance : falseDecisionPermitsTrueOther ? this : (Tests)Tests.False.Instance;
+                    Debug.Assert(sideeffect is True or One(BoundDagAssignmentEvaluation));
+                    // Given T0 as a test that has already occurred, T1 as a subsequent test, P as a pre-condition, and S as a side-effect (expected to always evaluate to True), we proceed as follow:
+                    //  - If T0=True implies T1, we rewrite T1 as (P ? S : T1) because we have determined that given T0=True, T1 would always succeed if P evaluates to True.
```

Comment:
>(P ? S : T1) [](http://example.com/codeflow?start=66&length=12)

I would find it extremely helpful if these were written in terms of ands and ors, rather than ternaries, as the rewritten sequences are actually ands and ors, not ternaries.